cou/steps/analyze.py

```diff
@@ -0,0 +1,488 @@
+    change_channel: defaultdict[str, set]
+    charmhub_migration: defaultdict[str, set]
+    change_openstack_release: defaultdict[str, set]
+
+
+@dataclass
+class Application:
+    """Representation of an application in the deployment."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    name: str
+    status: ApplicationStatus
+    config: dict
+    model_name: str
+    charm: str = ""
+    charm_origin: str = ""
+    os_origin: str = ""
+    channel: str = ""
+    pkg_name: str = ""
+    # E.g of units: {"keystone/0": {'os_version': 'victoria', 'pkg_version': '2:18.1'}}
+    units: defaultdict[str, dict] = field(default_factory=lambda: defaultdict(dict))
+    #  E.g of os_release_units: {"ussuri":{"keystone/0"}, "victoria": {"keystone/1"}}
+    os_release_units: defaultdict[str, set] = field(default_factory=lambda: defaultdict(set))
+    #  E.g of pkg_version_units: {"2:17.0": {"keystone/0"}, "2:18.0": {"keystone/1"}}
+    pkg_version_units: defaultdict[str, set] = field(default_factory=lambda: defaultdict(set))
+
+    def __post_init__(self) -> None:
+        """Initiate the Apllication dataclass."""
+        self.charm = self.extract_charm_name()
+        self.channel = self.status.charm_channel
+        self.charm_origin = self.status.charm.split(":")[0]
+        self.pkg_name = self.get_pkg_name()
+        self.os_origin = self.get_os_origin()
+        for unit in self.status.units.keys():
+            os_version = self.get_current_os_versions(unit)
+            self.units[unit]["os_version"] = os_version
+            self.os_release_units[os_version].add(unit)
+
+    def __hash__(self) -> int:
+        """Hash magic method for Application."""
+        return hash(self.name)
+
+    def __eq__(self, other: Any) -> Any:
+        """Equal magic method for Application."""
+        return other.name == self.name
+
+    def to_dict(self) -> Dict:
+        """Return a string in yaml format.
+
+        Passing the Application class directly to dump contain some fields that are big,
+        e.g: config and status. This output contains just the important fields for
+        the operator.
+        """
+        return {
+            self.name: {
+                "model_name": self.model_name,
+                "charm_origin": self.charm_origin,
+                "os_origin": self.os_origin,
+                "channel": self.channel,
+                "pkg_name": self.pkg_name,
+                "units": {
+                    unit: {
+                        "pkg_version": details.get("pkg_version", ""),
+                        "os_version": details.get("os_version", ""),
+                    }
+                    for unit, details in self.units.items()
+                },
+            }
+        }
+
+    def extract_charm_name(self) -> str:
+        """Extract the charm name using regex."""
+        match = re.match(
+            r"^(?:\w+:)?(?:~[\w\.-]+/)?(?:\w+/)*([a-zA-Z0-9-]+?)(?:-\d+)?$", self.status.charm
```

Comment:
has this been reverse engineered? do we have a formal and trusted specification of how a charm name is composed?

Reply:
It seems that comes from [here](https://github.com/juju/charm/blob/master/url.go#L51). What python libjuju shows as the field `charm` seems to be this url.